allennlp/data/dataset_readers/encoder_decoder.py

```diff
@@ -0,0 +1,91 @@
+from typing import Dict
+import logging
+
+from overrides import overrides
+
+import tqdm
+
+from allennlp.common import Params
+from allennlp.common.checks import ConfigurationError
+from allennlp.data.dataset import Dataset
+from allennlp.data.dataset_readers.dataset_reader import DatasetReader
+from allennlp.data.fields import TextField
+from allennlp.data.instance import Instance
+from allennlp.data.tokenizers import Tokenizer, WordTokenizer
+from allennlp.data.token_indexers import TokenIndexer, SingleIdTokenIndexer
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+START_SYMBOL = "@@START@@"
+END_SYMBOL = "@@END@@"
+
+@DatasetReader.register("encoder_decoder")
+class EncoderDecoderDatasetReader(DatasetReader):
+    """
+    Read a tsv file containing paired sequences, and create a dataset for an ``EncoderDecoder`` model.
+
+    Parameters
+    ----------
+    source_tokenizer : ``Tokenizer``, optional (default=``WordTokenizer()``)
+        Tokenizer to use to split the input sequences into words or other kinds of tokens.
+    target_tokenizer : ``Tokenizer``, optional (default=``WordTokenizer(start_tokens=[START_SYMBOL],
+                                                                        end_tokens=[END_SYMBOL])``)
+        Tokenizer to use to split the output sequences (during training) into words or other kinds of tokens.
+    source_token_indexers : ``Dict[str, TokenIndexer]``, optional
+                             (default=``{"tokens": SingleIdTokenizer("input_tokens")}``)
+        Indexers used to define input (source side) token representations.
+    target_token_indexers : ``Dict[str, TokenIndexer]``, optional
+                            (default=``{"tokens": SingleIdTokenizer("output_tokens")}``)
+        Indexers used to define output (target side) token representations.
+    """
+    def __init__(self, source_tokenizer: Tokenizer = None, target_tokenizer: Tokenizer = None,
+                 source_token_indexers: Dict[str, TokenIndexer] = None,
+                 target_token_indexers: Dict[str, TokenIndexer] = None) -> None:
+        self._source_tokenizer = source_tokenizer or WordTokenizer()
+        self._target_tokenizer = target_tokenizer or WordTokenizer(start_tokens=[START_SYMBOL],
+                                                                   end_tokens=[END_SYMBOL])
+        self._source_token_indexers = source_token_indexers or {"tokens": SingleIdTokenIndexer("source_tokens")}
+        self._target_token_indexers = target_token_indexers or {"tokens": SingleIdTokenIndexer("target_tokens")}
```

Comment:
@matt-gardner I believe you were unsure about the way I used namespaces here. Do you have better suggestions?
Once that is resolved, do you think this PR ready to be merged?

Reply:
I think it _might_ make sense to have the `"source_tokens"` and `"target_tokens"` strings be configurable.  But now that I see that the token indexers themselves are configurable, you might not need to do this.

I haven't looked at the code in detail yet, because it was marked [WIP], and you didn't request a review.  I'll review it today.

Reply:
Yeah, when I submitted this PR, I didn't have concrete plans for what's going in it. But you recently said that it may be a good idea to merge this and open a new PR for the constrained stuff, and I agree. So yeah, please review it when you get a chance.